File: ROS_Packages/src/maze_test/maze_game/scripts/GameModel.py
#!/usr/bin/env python

import rospkg
import Orange
from Orange.data import *
from Orange.feature import *
import rospy
from geometry_msgs.msg import Point
from visualization_msgs.msg import Marker
from move_base_msgs.msg import MoveBaseAction
from move_base_msgs.msg import MoveBaseGoal
import actionlib

import RunData
import Robot
from Robot import RobotState


class GameModel:

    def __init__(self):

        self.robots = {} # Holds the Robot objects for the robots we are representing in the game keyed by their name
        self.marker_pub = rospy.Publisher("visualization_marker", Marker, queue_size=10) # Publishes the maze as a Visualization Marker message, used primarily by rosbridge
        self.maze_points = Marker()

        # Data recording
        self.averageRunTimesVector = []# Holds the average times
        self.averageRunTime = rospy.Duration(0)
        self.averageHealthVector = []
        self.averageHealth = 0
        self.expectedPoints = [] # The points read using readPoints. They are used in the disparity calculation.

        self.readPoints()

        rospy.loginfo("GameModel ready!")

        self.runDataVector = [] #Stores all the RunData objects since the last export.

    def readPoints(self):
        print "Reading the points...\n"

        points_arr = []
        # path = rospkg.get_ros_package_path("maze_game") + "/include/points.txt"
        path = "/home/matt/cognitive_stress/ROS_Packages/src/maze_game/include/points.txt"
        print path + "\n"
        infile = open(path)

        for line in infile:

            line = line[:-1]
            split = line.split("  ")
            # Read the coords in as float (they must be space delimited!)
            x = float(split[0])
            y = float(split[1])
            z = float(split[2])

            p = Point

            p.x = x
            p.y = y
            p.z = z

            points_arr.append(p)

        # Set up the visualization markers
        rospy.loginfo("Building visualization markers...")
        self.maze_points.header.frame_id = "/map"
        self.maze_points.header.stamp = rospy.Time.now()
        self.maze_points.ns = "maze_walls"
        self.maze_points.action = Marker.ADD
        self.maze_points.pose.orientation.w = 1.0
        self.maze_points.id = 0
        self.maze_points.type = Marker.LINE_STRIP
        self.maze_points.scale.x = 0.05
        self.maze_points.scale.y = 0.05
        self.maze_points.color.r = 1.0
        self.maze_points.color.a = 1.0

        for i in range(len(points_arr)):
            self.expectedPoints.append(points_arr[i])
            self.maze_points.points.append(points_arr[i])

            if i < len(points_arr) - 1:
                self.maze_points.points.append(points_arr[i + 1])

    def addRobot(self, robotName):
        print "Adding robot: " + robotName + "\n"
        # Commented out for testing without creating new robot objects, DO NOT DELETE!

        toAdd = Robot.make_robot(robotName, self)
        self.robots.update(robotName, toAdd)

        # List robots
        i = 0
        print "---------- Robots List ----------\n"
        for key in self.robots:
            print("%i) %s\n", i, key)
            i += 1

    def removeRobot(self, robotName):
        print("Removing robot %s\n", robotName)

        self.robots.__delitem__(robotName)

        # List robots
        j = 0
        print "---------- Robots List ----------\n"
        for key in self.robots:
            print("%i) %s\n", j, key)
            j += 1

    def startRobot(self, robotName):
        ac = actionlib.SimpleActionClient("/" + robotName + "/move_base", MoveBaseAction)

        while not ac.wait_for_server(rospy.Duration(5)):
            rospy.loginfo("Waiting for move base action server...")

        self.robots[robotName].setState(RobotState.MOVING_TO_START)

        msg = MoveBaseGoal()

        msg.target_pose.header.frame_id = "map"
        msg.target_pose.header.stamp = rospy.Time.now()

        msg.target_pose.pose.position.x = 6.730
        msg.target_pose.pose.position.y =  6.0
        msg.target_pose.pose.position.z  = 0
        msg.target_pose.pose.orientation.x = 0
        msg.target_pose.pose.orientation.y  = 0
        msg.target_pose.pose.orientation.z = .993607
        msg.target_pose.pose.orientation.w = -0.112894

        rospy.loginfo("Sending %s to start of the maze...", robotName)
        ac.send_goal(msg)
        ac.stop_tracking_goal()

    def restartRobot(self, robot_name):
        # Check for the robots existence

        if self.robots.has_key(robot_name):
            self.robots[robot_name].resetRobot()
            rospy.loginfo("Successfully reset %s", robot_name)
        else:
            rospy.logwarn("No robot names %s to reset. ", robot_name)

    def checkRobotsPosition(self):

    # rospy.loginfo("Checking robots from GameModel.")

        for key, value in self.robots.iteritems():

        # self is not executed, seems to crash while adding a new robot...
        # rospy.loginfo("Checking robot %s", entry.first)

            if value.getCurState() == RobotState.MOVING_TO_START or value.getCurState() == RobotState.RUNNING_MAZE:
                # rospy.loginfo("%s has started, checking distance.", entry.first)
                value.checkDistance()

    def getExpPointAt(self, i):
        return self.expectedPoints[i]

    def getAverageScores(self):
        return [self.averageRunTime, self.averageHealth]

    def submitAndUpdateAverageScores(self, name, times, healthPoints, givenPoints, expectedPoints, inputDelay, trustScores):
        self.runDataVector.append(RunData.make_run_data(name, False, times, healthPoints, givenPoints, expectedPoints, inputDelay, trustScores))

        # Compute average duration
        self.averageRunTimesVector.append(rospy.Duration(times[-1:] - times[0]))
        self.averageRunTime = rospy.Duration((self.averageRunTime + self.averageRunTimesVector[-1:]).toSec() / len(self.averageRunTimesVector))

        # Compute average health
        self.averageHealthVector.append(healthPoints[-1:] - healthPoints.front())

    def exportRunData(self, robotName):

        # Vector to contain the robot(s) that we are asked to export data for
        toExport = []

        rospy.loginfo("Exporting data for %s...", robotName)

        # Check to see what we need to export
        if not robotName == "ALL":

            # Check for the robots existence (I realize this is redundant).

            if not len(self.runDataVector) == 0:
                for dat in self.runDataVector:
                    if dat.robotName == robotName:
                        toExport.append(dat)
        else:
            toExport = self.runDataVector

        for entry in toExport:

            path = rospkg.get_ros_package_path("maze_game")
            path += "/export/" + entry.timestamp + "_" + entry.robotName + ".tab"

            # Normalize the trust values

            maxTrust = max(entry.trustScores)
            minTrust = min(entry.trustScores)

            for i in range(len(entry.trustScores)):
                entry.trustScores[i] = 1 - (entry.trustScores[i] - minTrust) / (maxTrust - minTrust)

            Name, RunTime, RunBefore, GivenX, GivenY, ExpectedX, ExpectedY, Disparity, DamagePercent, TimeDelay, TrustValue = [Orange.feature.Discreate(x) for x in ['name', 'runtime', 'run before', 'given x', 'given y', 'expected x', 'expected y', 'disparity', 'percent damage', 'time delay', 'trust value']]
            dom = Domain([Name, RunTime, RunBefore, GivenX, GivenY, ExpectedX, ExpectedY, Disparity, DamagePercent, TimeDelay, TrustValue])
            data = Table(dom, [entry.robotName, entry.runMazeBefore, [p.x for p in entry.givenPoints], [p.y for p in entry.givenPoints], [p.x for p in entry.expectedPoints], [p.y for p in entry.expectedPoints], entry.disparity, entry.healthAtGoal, entry.inputDelay, entry.trustScores])

            print data

            data.save(path)

        # Clear the runData we just exported?
        return True

    def publishMazeMarkers(self):
        self.marker_pub.publish(self.maze_points)

    def getRobot(self, name):
        return self.robots[name]

def make_model():
    return GameModel()

File: ROS_Packages/src/maze_game/src/GameModel.cpp
#include "GameModel.h"
#include "Robot.h"

GameModel::GameModel(ros::NodeHandle *n) {
	this->nh = *n;
	readPoints();
	readExpectedArrivalTimes();

	ROS_INFO("GameModel ready!");
}

void GameModel::readPoints() {
	std::cout << "Reading the points...\n";

	std::vector<geometry_msgs::Point> points_arr;

	std::ifstream infile;
	std::string path = ros::package::getPath("maze_game").append("/include/points.txt");
	std::cout << path.c_str() << "\n";
	infile.open(path.c_str());

	if (!infile.is_open()) {
		std::cerr << "Error! Could not open points file";
		return;
	}

	std::string line;

	while (std::getline(infile, line)) {

		std::stringstream ss(line);

		float x, y, z;

		//Read the coords in as float (they must be space delimited!)
		ss >> x >> y >> z;

		geometry_msgs::Point p;

		p.x = x;
		p.y = y;
		p.z = z;

		points_arr.push_back(p);
	}

	//Set up the visualization markers
	ROS_INFO("Building visualization markers...");
	this->marker_pub = nh.advertise<visualization_msgs::Marker>("visualization_marker", 10);
	this->maze_points.header.frame_id = "/map";
	this->maze_points.header.stamp = ros::Time::now();
	this->maze_points.ns = "maze_walls";
	this->maze_points.action = visualization_msgs::Marker::ADD;
	this->maze_points.pose.orientation.w = 1.0;
	this->maze_points.id = 0;
	this->maze_points.type = visualization_msgs::Marker::LINE_STRIP;
	this->maze_points.scale.x = 0.05;
	this->maze_points.scale.y = 0.05;
	this->maze_points.color.r = 1.0;
	this->maze_points.color.a = 1.0;

	for (uint32_t i = 0; i < points_arr.size(); i++) {
		expectedPoints.push_back(points_arr.at(i));
		this->maze_points.points.push_back(points_arr.at(i));

		if (i < points_arr.size() - 1) {
			this->maze_points.points.push_back(points_arr.at(i + 1));
		}
	}
}

void GameModel::readExpectedArrivalTimes() {
	ROS_INFO("Reading the arrival Times...");

	std::ifstream infile;
	std::string path = ros::package::getPath("maze_game").append("/include/expectedTimes.txt");
	std::cout << path.c_str() << "\n";
	infile.open(path.c_str());

	if (!infile.is_open()) {
		std::cerr << "Error! Could not open expected times file";
		return;
	}

	std::string line;

	while (std::getline(infile, line)) {

		std::stringstream ss(line);

		double time;

		//Read the coords in as float (they must be space delimited!)
		ss >> time;

		this->expectedArrivalTimes.push_back(time);
	}
}


void GameModel::addRobot(std::string name) {
	printf("Adding robot: %s\n", name.c_str());
// Commented out for testing without creating new robot objects, DO NOT DELETE!

	Robot* toAdd = new Robot(name, &nh, this);
	robots.insert(std::make_pair(name, *toAdd));

	//List robots
	int i = 0;
	std::cout << "---------- Robots List ----------\n";
	for(auto &entry : this->robots) {
		printf("%i) %s\n", i, entry.first.c_str());
		i++;
	}
}

void GameModel::removeRobot(std::string name) {
	printf("Removing robot %s\n", name.c_str());

	this->robots.erase(name);

	//List robots
	int j = 0;
	std::cout << "---------- Robots List ----------\n";
	for(auto &entry : this->robots) {
		printf("%i) %s\n", j, entry.first.c_str());
		j++;
	}
}

void GameModel::startRobot(std::string robotName) {


	MoveBaseClient ac("/" + robotName + "/move_base", true);

	while(!ac.waitForServer(ros::Duration(5.0))) {
		ROS_INFO("Waiting for move base action server...");
	}

	this->robots.at(robotName).setState(RobotState::MOVING_TO_START);

	move_base_msgs::MoveBaseGoal msg;

	msg.target_pose.header.frame_id = "map";
	msg.target_pose.header.stamp = ros::Time::now();

	msg.target_pose.pose.position.x = 6.730;
	msg.target_pose.pose.position.y =  6.0;
	msg.target_pose.pose.position.z  = 0;
	msg.target_pose.pose.orientation.x = 0;
	msg.target_pose.pose.orientation.y  = 0;
	msg.target_pose.pose.orientation.z = .993607;
	msg.target_pose.pose.orientation.w = -0.112894;

	ROS_INFO("Sending %s to start of the maze...", robotName.c_str());
	ac.sendGoal(msg);
	ac.stopTrackingGoal();
}

void GameModel::restartRobot(std::string robot_name) {

	//Check for the robots existence
	std::map<std::string, Robot>::iterator it = this->robots.find(robot_name);

	if (it != this->robots.end()) {
		this->robots.at(robot_name).resetRobot();
		ROS_INFO("Successfully reset %s", robot_name.c_str());
	} else {
		ROS_WARN("No robot names %s to reset. ", robot_name.c_str());
	}
}

Robot* GameModel::getRobot(std::string name) {

	return &robots.at(name);
}

void GameModel::checkRobotsPosition() {

//	ROS_INFO("Checking robots from GameModel.");

	for (auto &entry : this->robots) {

		//This is not executed, seems to crash while adding a new robot...
//		ROS_INFO("Checking robot %s", entry.first.c_str());

		if (entry.second.getCurState() == RobotState::MOVING_TO_START || entry.second.getCurState() == RobotState::RUNNING_MAZE) {
//			ROS_INFO("%s has started, checking distance.", entry.first.c_str());
			entry.second.checkDistance();
		}
	}
}

geometry_msgs::Point GameModel::getExpPointAt(int i) {
	return expectedPoints.at(i);
}

std::pair<ros::Duration, int32_t> GameModel::getAverageScores() {
	std::pair<ros::Duration, int32_t> toReturn = std::make_pair(this->averageRunTime, this->averageHealth);
	return toReturn;
}

void GameModel::submitAndUpdateAverageScores(std::string name, std::vector<double> times, std::vector<int32_t> healthPoints, std::vector<double> percentDamage, std::vector<geometry_msgs::Point> givenPoints, std::vector<geometry_msgs::Point> expectedPoints, std::vector<geometry_msgs::Point> disparity, std::vector<double> inputDelay, std::vector<double> trustScores, std::map<int, std::vector<double>> goalFrequencyForPoints) {

	//Record run completion
	this->runDataVector.push_back(RunData(name, false, times, healthPoints, percentDamage, givenPoints, expectedPoints, disparity, inputDelay, trustScores, goalFrequencyForPoints));

	//Compute average duration
	this->averageRunTimesVector.push_back(ros::Duration(times.back() - times.front()));
	this->averageRunTime = ros::Duration((this->averageRunTime + this->averageRunTimesVector.back()).toSec() / this->averageRunTimesVector.size());

	//Compute average health
	this->averageHealthVector.push_back(healthPoints.back() - healthPoints.front());
	this->averageHealth = (this->averageHealth + this->averageHealthVector.back()) / this->averageHealthVector.size();
}

std::string processDoubleValue(double val) {

	std::stringstream ss;

	if (val == -86401.0) {
		return std::string("*");
	} else {
		ss << val;
		return ss.str();
	}
}

void GameModel::computeValuesForDataSample(GameModel::RunData& data) {

	for (int i = data.disparity.size() - 1; i >= 0; i--) {

		//Fix first row
		if (i == 0) {
			data.goalArivalTimes[i] -= data.goalArivalTimes[0];
			data.percentDamage[i] = 0.0;
			data.goalArrivalDelay[i] = 0.0;
			break;
		}

		//Compute arrival times
		data.goalArivalTimes[i] -= data.goalArivalTimes[0];
		printf("Comp. goalArrivalTime @ %i: %.6f\n", i, data.goalArivalTimes[i]);

		//Compute Percent Damage Values
		data.percentDamage[i] = double((data.healthAtGoal[0] - data.healthAtGoal[i]) / data.healthAtGoal[0]);
		printf("Comp. percentDamage @ %i: %.6f\n", i, data.percentDamage[i]);

		//Compute Goal arrival delay
		data.goalArrivalDelay[i] = this->expectedArrivalTimes[i] - data.goalArivalTimes[i];
		printf("Comp. goalArrivalDelay @ %i: %.6f\n", i, data.goalArrivalDelay[i]);

		//Compute trust score (unnormalized trust value)
		data.trustScores[i] = data.disparity[i].z + data.percentDamage[i] + data.goalArrivalDelay[i];
		printf("Comp. trustScore @ %i: %.6f\n", i, data.trustScores[i]);
	}
}

bool GameModel::exportRunData(std::string robotName) {

	//Vector to contain the robot(s) that we are asked to export data for
	std::vector<RunData> toExport;

	ROS_INFO("Exporting data for %s...", robotName.c_str());

	//Check to see what we need to export
	if (robotName != "ALL") {

		//Check for the robots existence (this is redundant I realize).
		std::vector<RunData>::iterator it = std::find_if(this->runDataVector.begin(), this->runDataVector.end(),
				[&robotName](const RunData& data) {
					return data.robotName.compare(robotName) == 0;
				});

		if (it != this->runDataVector.end()) {
			for (RunData dat : this->runDataVector) {
				if (dat.robotName.compare(robotName) == 0) {
					toExport.push_back(dat);
				}
			}
		}
	} else {
		toExport = this->runDataVector;
	}

	//File to export to
	std::ofstream outFile;

	for (RunData data : toExport) {

		computeValuesForDataSample(data);

		std::string path = ros::package::getPath("maze_game");
		path.append(std::string("/export/" + std::to_string(data.timestamp) + "_" + data.robotName + ".csv"));
		outFile.open(path.c_str(), std::ios_base::app);

		if (!outFile.is_open()) {
			ROS_ERROR("Could not open output path at: %s", path.c_str());
			return false;
		}

		std::string csvToExport = "Robot,Runtime,Run Before,Arrival Time,Health at Goal,Percent Damage,Given x,Given y,Expected x,Expected y,Disparity,Goal Arrival Delay,Trust Value\n";

		//Normalize the trust values (start at index 1 since we do not wish to include the first scores as it will always be 1)
		double maxTrust = data.trustScores[1];
		double minTrust = data.trustScores[1];

		for (double e : data.trustScores) {
			if (e != -86401.0) {

				if (e > maxTrust) {
					maxTrust = e;
				} else if (e < minTrust) {
					minTrust = e;
				}

			}
		}

//		maxTrust = *std::max_element(data.trustScores.begin(), data.trustScores.end());
//		minTrust = *std::min_element(data.trustScores.begin(), data.trustScores.end());

		ROS_INFO("maxTrust: %.6f, minTrust: %.6f", maxTrust, minTrust);

		for (int i = 0; i < data.trustScores.size(); i++) {
			data.trustScores[i] = ((data.trustScores[i] - minTrust)/ (maxTrust - minTrust));
		}

		for (int i = 0; i < data.trustScores.size(); i++) {

			std::string curRow = "";

			//Some values only have actual data on the first row
			if (i == 0) {
				curRow.append(data.robotName
							  + ","
							  + std::to_string(data.goalArivalTimes.back() - data.goalArivalTimes.front())
							  + ","
							  + (data.runMazeBefore ? "True" : "False")
							  + ","
						);
			} else {
				curRow.append("*,*,*,");
			}

			curRow.append(processDoubleValue(data.goalArivalTimes[i]) + ",");

			if (data.healthAtGoal[i] == INT32_MIN) {
				curRow.append("*,");
			} else {
				curRow.append(std::to_string(data.healthAtGoal[i]) + ",");
			}

			curRow.append(processDoubleValue(data.percentDamage[i]) + ",");
			curRow.append(processDoubleValue(data.givenPoints[i].x) + ",");
			curRow.append(processDoubleValue(data.givenPoints[i].y) + ",");
			curRow.append(processDoubleValue(data.expectedPoints[i].x) + ",");
			curRow.append(processDoubleValue(data.expectedPoints[i].y) + ",");
			curRow.append(processDoubleValue(data.disparity[i].z) + ",");
			curRow.append(processDoubleValue(data.goalArrivalDelay[i]) + ",");
			curRow.append(processDoubleValue(data.trustScores[i]) + "\n");

			csvToExport.append(curRow);
		}

		outFile << csvToExport;

		ROS_INFO("Finished export for %s!", robotName.c_str());
	}

	//Clear the runData we just exported?

	return true;
}

void GameModel::publishMazeMarkers() {

	this->marker_pub.publish(this->maze_points);
}
